### DISCUSSION TITLE
Steady advection problem

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#23534] [Q&A Modules: General] [answered]
Hi!
I am trying to solve a simple 1D problem comprising of advection and source term:
\partial{phi}/\partial{x}=10

I am expecting results as a straight line phi = 10x

But my input file doesn't converges. Please have a look at my input file. I am using navierstokes module.

Thanks

`[Mesh]
  type = GeneratedMesh
  dim = 1
  xmin = 0
  xmax = 10.0
  nx = 100
[]

[Variables]
  [./phi]
  [../]
[]

[Kernels]
  [./advection]
    type = MassConvectiveFlux
    variable = phi
    vel_x = 1
  [../]

  [./rhs]
    type = CoupledForce
    variable = phi
    v = 10
  [../]
[]

[BCs]
  [./left]
    type = DirichletBC
    variable = phi
    boundary = 'left'
    value = 0
  [../]
[]

[Executioner]
  type = Steady
[]
[Outputs]
  exodus = true
[]
`

**Comment by jiangwen84** [comment, marked as answer]
This is a purely advection equation. You probably need some sort of stabilization, e.g., SUPG.

There are several similar implementation of advection equations and SUPG stabilization in different physics modules. Navier Stokes modules have those implemented, but not for this simple equation. Instead, I used level set module to solve your problem. If you do not add SUPG, the solver will fail due to ```FACTOR_NUMERIC_ZEROPIVOT```. Adding the SUPG term will make it solve. 

```
[Mesh]
  type = GeneratedMesh
  dim = 1
  xmin = 0
  xmax = 10.0
  nx = 10
[]

[Variables]
  [phi]
  []
[]

[AuxVariables]
  [velocity]
    family = LAGRANGE_VEC
  []
[]

[AuxKernels]
  [vec]
    type = VectorFunctionAux
    variable = velocity
    function = velocity_func
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]

[Functions]
  [velocity_func]
    type = ParsedVectorFunction
    expression_x = '1'
  []
[]

[Kernels]
  [advection]
    type = LevelSetAdvection
    velocity = velocity
    variable = phi
  []

  [advection_supg]
    type = LevelSetAdvectionSUPG
    velocity = velocity
    variable = phi
  []

  [rhs]
    type = BodyForce
    variable = phi
    value = 10
  []
[]

[BCs]
  [left]
    type = DirichletBC
    variable = phi
    boundary = 'left'
    value = 0
  []
[]

[Executioner]
  type = Steady
[]
[Outputs]
  exodus = true
[]
```

**Reply by vermaprk** [reply]
Thanks. I tried to run the input file. But "FACTOR_NUMERIC_ZEROPIVOT" is still there.
I have a time dependent PDE in which these terms are present. I am doing this to verify whether each terms are giving the required output. Is there any other way to test individual terms ?
Thanks again!

`*** Warning ***
/Users/prakharverma/projects/newmanmodel/problem/levelset00.i:32: unused parameter 'Functions/velocity_func/expression_x'

 0 Nonlinear |R| = 3.041381e+00
      0 Linear |R| = 3.041381e+00
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
Nonlinear solve did not converge due to DIVERGED_LOCAL_MIN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge`

**Reply by GiudGiud** [reply]
This unused param makes me think you are using an old version of MOOSE where the parameter had a different name

Please update

Guillaume

**Reply by vermaprk** [reply]
Thanks. Updated to latest version. Its working fine now.
However, my other question was:
I have a PDE in which these terms are present. I am doing this to verify whether each terms individually are giving the required output. Is there any other way to test individual terms ?

**Reply by GiudGiud** [reply]
There s the residual tagging method and you can lead method of manufactured solutions studies pretty easily with our utility script

there s good documentation on the website if you want to check it out

**Reply by GiudGiud** [reply]
On tagging
https://mooseframework.inl.gov/framework_development/tagging.html
then you use a TaggingAux
https://mooseframework.inl.gov/moose/source/auxkernels/TagMatrixAux.html

On MMS
https://mooseframework.inl.gov/python/mms.html
